src/emcee.jl

```diff
@@ -0,0 +1,223 @@
+struct Ensemble{D} <: MHSampler
+    n_walkers::Int
+    proposal::D
+end
+
+# Define the first step! function, which is called at the 
+# beginning of sampling. Return the initial parameter used
+# to define the sampler.
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::Ensemble,
+    N::Integer,
+    ::Nothing;
+    init_params = nothing,
+    kwargs...,
+)
+    if init_params === nothing
+        return propose(rng, spl, model)
+    else
+        return Transition(model, init_params)
+    end
+end
+
+# Define the other step functions. Returns a Transition containing
+# either a new proposal (if accepted) or the previous proposal 
+# (if not accepted).
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::Ensemble,
+    ::Integer,
+    params_prev;
+    kwargs...,
+)
+    # Generate a new proposal. Accept/reject happens at proposal level.
+    return propose(rng, spl, model, params_prev)
+end
+
+#
+# Initial proposal
+# 
+function propose(rng::Random.AbstractRNG, spl::Ensemble, model::DensityModel)
+    # Make the first proposal with a static draw from the prior.
+    static_prop = StaticProposal(spl.proposal.proposal)
+    mh_spl = MetropolisHastings(static_prop)
+    return [propose(rng, mh_spl, model) for _ in 1:spl.n_walkers]
+end
+
+
+
+#
+# Every other proposal
+# 
+function propose(rng::Random.AbstractRNG, spl::Ensemble, model::DensityModel, walkers::Vector{W}) where {W<:Transition}
+    new_walkers = Vector{W}(undef, spl.n_walkers)
+    interval = 1:spl.n_walkers
+
+
+    nwalkers = spl.n_walkers
+    others = 1:(nwalkers - 1)
+    for i in interval
+        walker = walkers[i]
+        idx = mod1(i + rand(rng, others), nwalkers)
+        other_walker = walkers[idx]
+        new_walkers[i] = move(rng, spl, model, walker, other_walker)
+    end
+
+    return new_walkers
+end
+
+
+#####################################
+# Basic stretch move implementation #
+#####################################
+struct StretchProposal{P, F<:AbstractFloat} <: Proposal{P}
+    proposal :: P
+    stretch_length::F
+end
+
+StretchProposal(p) = StretchProposal(p, 2.0)
+
+function move(
+    rng::Random.AbstractRNG, 
+    spl::Ensemble{<:StretchProposal},
+    model::DensityModel,
+    walker::Transition,
+    other_walker::Transition,
+)
+    # Calculate intermediate values
+    proposal = spl.proposal
+    n = length(walker.params)
+    a = proposal.stretch_length
+    z = ((a - 1) * rand(rng) + 1)^2 / a
+    alphamult = (n - 1) * log(z)
+
+    # Make new parameters
+    y = walker.params + z .* (other_walker.params - walker.params)
```

Comment:
Maybe broadcast everything here?
```suggestion
    y = @. walker.params + z * (other_walker.params - walker.params)
```